File: backend/games/models/game.py
from django.core.validators import MaxValueValidator, MinValueValidator
from django.db import models

class Game(models.Model):
    class Status(models.TextChoices):
        SETUP = "setup", "Setup"
        LIVE = "live", "Live"
        FINAL = "final", "Final"

    class HalfInning(models.TextChoices):
        TOP = "top", "Top"
        BOTTOM = "bottom", "Bottom"

    status = models.CharField(
        max_length=20,
        choices=Status.choices,
        default=Status.SETUP,
    )

    scheduled_innings = models.PositiveIntegerField(
        default=7,
        validators=[
            MinValueValidator(1),
            MaxValueValidator(20),
        ],
    )

    inning = models.PositiveSmallIntegerField(default=1)

    half_inning = models.CharField(
        max_length=10,
        choices=HalfInning.choices,
        default=HalfInning.TOP,
    )

    outs = models.PositiveSmallIntegerField(default=0)

    home_score = models.PositiveIntegerField(default=0)
    away_score = models.PositiveIntegerField(default=0)

    current_home_batter_index = models.PositiveIntegerField(default=0)
    current_away_batter_index = models.PositiveIntegerField(default=0)

    runner_on_first = models.ForeignKey(
        "games.GamePlayer",
        null=True,
        blank=True,
        related_name="games_as_runner_on_first",
        on_delete=models.SET_NULL,
    )

    runner_on_second = models.ForeignKey(
        "games.GamePlayer",
        null=True,
        blank=True,
        related_name="games_as_runner_on_second",
        on_delete=models.SET_NULL,
    )

    runner_on_third = models.ForeignKey(
        "games.GamePlayer",
        null=True,
        blank=True,
        related_name="games_as_runner_on_third",
        on_delete=models.SET_NULL,
    )

    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    @property
    def batting_side(self):
        if self.half_inning == self.HalfInning.TOP:
            return "away"

        return "home"

    def __str__(self):
        return f"Game #{self.id}"

File: frontend/src/types/game.ts
export type GameStatus = 'setup' | 'live' | 'final'
export type HalfInning = 'top' | 'bottom'
export type TeamSide = 'home' | 'away'

export type PlateAppearanceResult =
  | 'single'
  | 'double'
  | 'triple'
  | 'home_run'
  | 'walk'
  | 'strikeout'
  | 'ground_out'
  | 'fly_out'
  | 'error'
  | 'fielders_choice'

export type GamePlayer = {
  id: number
  name: string
  batting_category: string
  lineup_position: number | null
  is_active: boolean
}

export type GameTeam = {
  id: number
  side: TeamSide
  name: string
  batting_order_mode: string
  lineup_rule_config: Record<string, unknown>
  players: GamePlayer[]
}

export type Game = {
  id: number
  status: GameStatus
  scheduled_innings: number
  inning: number
  half_inning: HalfInning
  outs: number
  home_score: number
  away_score: number
  current_home_batter_index: number
  current_away_batter_index: number
  runner_on_first: number | null
  runner_on_first_name?: string
  runner_on_second: number | null
  runner_on_second_name?: string
  runner_on_third: number | null
  runner_on_third_name?: string
  batting_side: TeamSide
  current_batter: GamePlayer | null
  teams: GameTeam[]
  created_at: string
  updated_at: string
}

export type CreateGameInput = {
  home_team_name: string
  away_team_name: string
  scheduled_innings: number
  home_lineup: Array<{ name: string }>
  away_lineup: Array<{ name: string }>
}

export type PlateAppearanceResponse = {
  game: Game
  event_id: number
}
